### PR TITLE
Non vanilla world skip

### PR DESCRIPTION
**Pull request by Foghrye4**
Now after player connected to server it throw interesting exceptions and shut itself down.
See also #223

### CODE REVIEW

File: src/main/java/cubicchunks/CommonEventHandler.java [hunk 1]
```diff
@@ -133,11 +134,11 @@ public void onPlayerJoinWorld(EntityJoinWorldEvent evt) {
...
-        @Override public boolean test(ICubicWorld world) {
-            if (super.test(world)) {
-                return true;
-            }
-            if (((World) world).getChunkProvider().getClass() != ChunkProviderClient.class) {
+        @Override
+        public boolean test(ICubicWorld world) {
+            if (!super.test(world))
+                return false;
+            if (world.getClass() == (Class<?>) WorldClient.class && ((World) world).getChunkProvider().getClass() != ChunkProviderClient.class) {
```
Comment (Niko-sk2x) [inline src/main/java/cubicchunks/CommonEventHandler.java]: Why is that needed? If the chunk provider is of the wrong type, it doesn't matter if it's WorldClient or not

**Review by Niko-sk2x — COMMENTED** [review]
Seems like I made enough mistakes in there that it's surprising it worked in singleplayer. This seems to look good.

File: src/main/java/cubicchunks/CommonEventHandler.java (hunk 1 above)
Comment (Foghrye4) [reply src/main/java/cubicchunks/CommonEventHandler.java]: My thoughts: We load both WorldServer and WorldClient for a single player game ("Client" side).
Therefore there should be ChunkProviderServer as well, isn't it?

Comment (Niko-sk2x) [reply src/main/java/cubicchunks/CommonEventHandler.java]: i think we should just have a hardcoded list of allowed classes in each proxy, and we would just check if the list contains the class. Much less error-prone.

It would need list of allowed:
 * World classes implementations
 * Chunk privider implementations
 * anything more?

Comment (Foghrye4) [reply src/main/java/cubicchunks/CommonEventHandler.java]: Should we check ChunkProviders at server side?

Comment (Niko-sk2x) [reply src/main/java/cubicchunks/CommonEventHandler.java]: Yes

File: src/main/java/cubicchunks/CommonEventHandler.java [hunk 2]
```diff
@@ -132,24 +142,52 @@ public void onPlayerJoinWorld(EntityJoinWorldEvent evt) {
...
+                WorldClient.class,
+                WorldServer.class,
+                WorldServerMulti.class
+        });
+        
+        private final List<Class<?>> allowedClientChunkProviderClasses = ImmutableList.copyOf(new Class[] {
+                ChunkProviderClient.class,
+                ChunkProviderEnd.class,
+                ChunkProviderHell.class,
+                ChunkProviderFlat.class,
```
Comment (Niko-sk2x) [inline src/main/java/cubicchunks/CommonEventHandler.java]: WorldGenerators don't count as chunk privoders. Also make it `List<Class<? extends IChunkProvider>>`

**Comment by Niko-sk2x** [comment]
That looks right, merging it.
